pkgs/native_assets_cli/lib/src/api/build_output.dart

```diff
@@ -4,48 +4,123 @@
+  /// the build will be re-run. Typically these dependencies contain the
+  /// `build.dart` script itself, and the source files used in the build.
+  ///
+  /// Metadata can be passed to `build.dart` invocations of dependent packages.
+  /// It must be provided to the constructor as [metadata], or added later with
+  /// [addMetadatum] and [addMetadata].
   factory BuildOutput({
     DateTime? timestamp,
-    List<Asset>? assets,
-    Dependencies? dependencies,
-    Metadata? metadata,
+    Iterable<Asset>? assets,
+    Iterable<Uri>? dependencies,
+    Map<String, Object>? metadata,
   }) =>
-      model.BuildOutput(
+      BuildOutputImpl(
         timestamp: timestamp,
-        assets: assets?.map((e) => e as model.Asset).toList(),
-        dependencies: dependencies as model.Dependencies?,
-        metadata: metadata as model.Metadata?,
+        assets: assets?.cast<AssetImpl>().toList(),
+        dependencies: Dependencies([...?dependencies]),
+        metadata: Metadata({...?metadata}),
       );
 
+  /// Adds [Asset]s produced by this build or dry run.
+  ///
+  /// In dry runs, the [Architecture] for [CCodeAsset]s can be omitted.
+  void addAsset(Asset asset);
+
+  /// Adds [Asset]s produced by this build or dry run.
+  ///
+  /// In dry runs, the [Architecture] for [CCodeAsset]s can be omitted.
+  void addAssets(Iterable<Asset> assets);
+
+  /// Adds file used by this build.
+  ///
+  /// If any of the files are modified after [timestamp], the build will be
+  /// re-run.
+  void addDependency(Uri dependency);
+
+  /// Adds files used by this build.
+  ///
+  /// If any of the files are modified after [timestamp], the build will be
+  /// re-run.
+  void addDependencies(Iterable<Uri> dependencies);
+
+  /// Adds metadata to be passed to `build.dart` invocations of dependent
+  /// packages.
+  void addMetadatum(String key, Object value);
+
+  /// Adds metadata to be passed to `build.dart` invocations of dependent
+  /// packages.
+  void addMetadata(Map<String, Object> metadata);
+
   /// The version of [BuildOutput].
   ///
-  /// This class is used in the protocol between the Dart and Flutter SDKs
+  /// The build output is used in the protocol between the Dart and Flutter SDKs
   /// and packages through `build.dart` invocations.
-  ///
-  /// If we ever were to make breaking changes, it would be useful to give
-  /// proper error messages rather than just fail to parse the YAML
-  /// representation in the protocol.
-  static Version get version => model.BuildOutput.version;
+  static Version get latestVersion => BuildOutputImpl.latestVersion;
 
-  Future<void> writeToFile({required Uri outDir});
+  /// Write out this build output to a file inside [BuildConfig.outDir].
+  ///
+  /// This takes into account the [BuildConfig.latestVersion] to write the
+  /// build output in a format the SDK that invoked the script supports.
+  Future<void> writeToFile({required BuildConfig config});
```

Comment:
Should this be part of the API, or is this confusing given the new way of running a `build.dart` script with the callback?